src/runtime/contrib/libxsmm/libxsmm_json_runtime.cc

```diff
@@ -0,0 +1,192 @@
+        int ldc = N_;
+
+        // Curently we support fp32 only.
+        libxsmm_datatype dtype = LIBXSMM_DATATYPE_F32;
+
+        // Configure GEMM related parameters
+        libxsmm_bitfield l_flags = LIBXSMM_GEMM_FLAG_NONE | LIBXSMM_GEMM_FLAG_BETA_0;
+        libxsmm_bitfield l_prefetch_flags = LIBXSMM_GEMM_PREFETCH_NONE;
+        libxsmm_gemm_shape l_shape =
+            libxsmm_create_gemm_shape(N_, M_, K_, lda, ldb, ldc, dtype, dtype, dtype, dtype);
+        libxsmm_blasint stride_a = N_ * K_ * sizeof(float);
+        libxsmm_blasint stride_b = K_ * M_ * sizeof(float);
+        libxsmm_gemm_batch_reduce_config l_brconfig = libxsmm_create_gemm_batch_reduce_config(
+            LIBXSMM_GEMM_BATCH_REDUCE_STRIDE, stride_a, stride_b, 0 /*br_unrool_hint*/);
+
+        libxsmm_gemm_ext_unary_argops l_argops;
+        libxsmm_gemm_ext_binary_postops l_postops;
+        memset(&l_argops, 0, sizeof(libxsmm_gemm_ext_unary_argops));
+        memset(&l_postops, 0, sizeof(libxsmm_gemm_ext_binary_postops));
+
+        if (has_bias_) {
+          l_postops.d_in_type = dtype;
+          l_postops.d_binary_flags = LIBXSMM_MELTW_FLAG_BINARY_BCAST_COL_IN_0;
+          l_postops.d_binary_type = LIBXSMM_MELTW_TYPE_BINARY_ADD;
+          l_postops.ldd = ldc;
+        }
+
+        if (has_relu_) {
+          l_argops.cp_unary_flags = LIBXSMM_MELTW_FLAG_UNARY_NONE;
+          l_argops.cp_unary_type = LIBXSMM_MELTW_TYPE_UNARY_RELU;
+          l_argops.ldcp = ldc;
+          // relu mask should have the same size as matrix C.
+          relu_mask_.resize(M_ * N_, 0);
+        }
+
+        // Use "libxsmm_gemmfunction" for GEMM kernel, and "libxsmm_gemmfunction_ext" for fused GEMM
+        // kernel.
+        if (has_bias_ || has_relu_) {
+          gemm_fusion_kernel_ = libxsmm_dispatch_brgemm_ext_v2(l_shape, l_flags, l_prefetch_flags,
+                                                               l_brconfig, l_argops, l_postops);
+        } else {
+          gemm_kernel_ = libxsmm_dispatch_brgemm_v2(l_shape, l_flags, l_prefetch_flags, l_brconfig);
+        }
+      }
+    }
+  }
+
+  void Run() override {
+    // Get input/output buffers.
+    auto data_eid = EntryID(input_nodes_[0], 0);
+    auto filter_eid = EntryID(input_nodes_[1], 0);
+    auto output_eid = EntryID(outputs_[0]);
+
+    void* data_handle = data_entry_[data_eid]->data;
+    void* filter_handle = data_entry_[filter_eid]->data;
+    void* output_handle = data_entry_[output_eid]->data;
+
+    // Transpose weight matrix since libxsmm only support GEMM rather than DENSE.
+    if (!transposed_filter_handle_) {
+      TVMDeviceAllocDataSpace(dev_, K_ * N_ * sizeof(float), kAllocAlignment, type_hint_,
+                              &transposed_filter_handle_);
+      for (int k = 0; k < K_; ++k) {
+        for (int n = 0; n < N_; ++n) {
+          static_cast<float*>(transposed_filter_handle_)[k * N_ + n] =
+              static_cast<float*>(filter_handle)[n * K_ + k];
+        }
+      }
+    }
```

Comment:
This should be done at the compile time because weights are constants.
Also, if you attempt to manage constant weights in your codegen, you should have a constant updater to remove the one in metadata module; otherwise you'll actually have two weights stored in the binary. You can search "constant_updater" to see the use cases.

Reply:
Do transpose at compile time does make sense. Should I write a custom pass to do that transform or there is already a pass I can leverage?

Reply:
I think it can be done in your codegen.

Reply:
Done.